src/services/pool/decorators/pool.decorator.ts

```diff
@@ -35,41 +35,73 @@ export class PoolDecorator {
     currency: FiatCurrency,
     tokens: TokenInfoMap
   ): Promise<Pool[]> {
+    const processedPools = this.pools.map(pool => {
+      const poolService = new this.poolServiceClass(pool);
+      poolService.removePreMintedBPT();
+      poolService.setTotalLiquidity(prices, currency, tokens);
+      poolService.setUnwrappedTokens();
+
+      return poolService.pool;
+    });
+
     const [
       protocolFeePercentage,
       gaugeBALAprs,
       gaugeRewardTokenAprs,
       poolSnapshots,
       rawOnchainDataMap
-    ] = await this.getData(prices, gauges, tokens);
+    ] = await this.getData(prices, gauges, tokens, processedPools);
 
-    const promises = this.pools.map(async pool => {
+    const promises = processedPools.map(async pool => {
       const poolSnapshot = poolSnapshots.find(p => p.id === pool.id);
       const poolService = new this.poolServiceClass(pool);
 
-      poolService.removePreMintedBPT();
-      await poolService.setLinearPools();
       poolService.setOnchainData(rawOnchainDataMap[pool.id], tokens);
-      poolService.setTotalLiquidity(prices, currency, tokens);
       poolService.setFeesSnapshot(poolSnapshot);
       poolService.setVolumeSnapshot(poolSnapshot);
-      poolService.setUnwrappedTokens();
-
-      await poolService.setAPR(
-        poolSnapshot,
-        prices,
-        currency,
-        protocolFeePercentage,
-        gaugeBALAprs[pool.id],
-        gaugeRewardTokenAprs[pool.id]
-      );
 
+      await Promise.all([
+        poolService.setLinearPools(),
+        poolService.setAPR(
+          poolSnapshot,
+          prices,
+          currency,
+          protocolFeePercentage,
+          gaugeBALAprs[pool.id],
+          gaugeRewardTokenAprs[pool.id]
+        )
+      ]);
       return poolService.pool;
     });
 
     return await Promise.all(promises);
   }
 
+  public async decorateSinglePool(
+    prices: TokenPrices,
+    currency: FiatCurrency,
+    tokens: TokenInfoMap
+  ): Promise<Pool> {
+    const poolMulticaller = new PoolMulticaller(this.pools);
+    const [poolSnapshots, rawOnchainDataMap] = await Promise.all([
+      this.getSnapshots(),
+      poolMulticaller.fetch()
+    ]);
+    const pool = this.pools[0];
+    const poolSnapshot = poolSnapshots.find(p => p.id === pool.id);
+    const poolService = new this.poolServiceClass(pool);
+
+    poolService.removePreMintedBPT();
+    await poolService.setLinearPools();
+    poolService.setOnchainData(rawOnchainDataMap[pool.id], tokens);
+    poolService.setTotalLiquidity(prices, currency, tokens);
+    poolService.setFeesSnapshot(poolSnapshot);
+    poolService.setVolumeSnapshot(poolSnapshot);
+    poolService.setUnwrappedTokens();
+
+    return poolService.pool;
+  }
```

Comment:
Why is this needed? Is it just decoration without APR? If it is, maybe we should just add a flag to the decorate function like `excludeAPR: boolean` or something. Then skip that part in the function. Having two functions that repeat 90% of the code is not very dry.

Reply:
added logic with `includeAPR` argument